### ISSUE TITLE
Encourage Initialization of Logging Early

### ISSUE THREAD

**Issue by dazinator** [dotnet/aspnetcore#19807]
### Encourage early initialization of logging /  ILoggerProvider from program.cs

The current templates, and the way logging is advocated, results in logging not being able to be used until after the DI container has been built and an ILogger can then be resolved from it. This is very late in the process and it means a whole swathe of startup issues can occur right from the entry point in program.cs and can go un-logged.

### Describe the solution you'd like
I propose changing the templates and host builder (if necessary) to encourage a different paradigm where an ILoggerProvider is created as soon as possible from the entry point in Program.cs. From there, it would be great if the HostBuilder itself could leverage that ILoggerProvider perhaps for it's own logging, and also perhaps to allow an ILogger to be injected into startup classes (something that can't be done today), and finally so that the ILoggerProvider that was created could be automatically included in the list of Providers when AddLogging() is eventually involved some while later to configure logging for DI.

### Additional context
I am using this paradigm myself but it's my own boiler plate I have to introduce to every new dotnet core application to ensure I always capture log output. I also think if the hostbuilder was more aware of this ILoggerProvider then perhaps there would be benefits in it also being able to log stuff.

**Comment by davidfowl** [comment]
> The current templates, and the way logging is advocated, results in logging not being able to be used until after the DI container has been built and an ILogger can then be resolved from it. This is very late in the process and it means a whole swathe of startup issues can occur right from the entry point in program.cs and can go un-logged.

This change is non-trivial because it also means that configuration needs to be setup early as well and cannot be part of the builder.

> I am using this paradigm myself but it's my own boiler plate I have to introduce to every new dotnet core application to ensure I always capture log output. I also think if the hostbuilder was more aware of this ILoggerProvider then perhaps there would be benefits in it also being able to log stuff.

I don't know how this would be different from template code. This change separates logging and configuration from the host builder which itself will introduce boiler plate. 

To flesh this out a bit more, here's a proposal:

```C#
using System;
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;

namespace MyApplication
{
    public class Program
    {
        public static int Main(string[] args)
        {
            using var configuration = new ConfigurationBuilder()
                                    .AddConfigurationDefaults() // Where does this method live
                                    .Build(); 

            using var loggerFactory = LoggerFactory.Create(builder =>
            {
                builder.AddLoggingDefaults(configuration); // Where does this method come live?
            });

            var logger = loggerFactory.CreateLogger<Program>();

            try
            {
                CreateHostBuilder(args, configuration, loggerFactory).Build().Run();
            }
            catch (Exception ex)
            {
                logger.LogError(ex, "An error occured when starting the application");
                return 1;
            }

            return 0;
        }

        public static IHostBuilder CreateHostBuilder(string[] args, IConfiguration configuration, ILoggerFactory loggerFactory) =>
            Host.CreateDefaultBuilder(args)
                .UseLoggerFactory(loggerFactory) // New method
                .ConfigureAppConfiguration(builder =>
                {
                    builder.AddConfiguration(configuration); // Chain the host configuration
                })
                .ConfigureWebHostDefaults(builder =>
                {
                    builder.UseStartup<Startup>();
                });
    }
}
```

A few things to note:
- The configuration options (such as the base path) need to happen up front in Program.cs.
- We need these helper methods to add a default set of loggers and configuration need to be defined somewhere that references all configuration and logging providers.
- The base path is normally configured on the webhost and is applied to the configuration builder and the webroot, that no longer works here since configuration happens before the hostbuilder.
- We need to update the tools that call CreateHostBuilder to pass an IConfiguration and an ILoggerFactory.
- Serilog needs to expose a factory for their ILoggerFactory instead of it being an extension method on the IHostBuilder.

cc @DamianEdwards

**Comment by dazinator** [comment]
> This change is non-trivial because it also means that configuration needs to be setup early as well and cannot be part of the builder.

Thanks for giving this some more thought David. I agree that initialising logging from config is very valuable, but for this initial logger, there is something to be said for minimising dependencies. The minimum that needs to happen as far as I can see is that an ILoggerProvider needs to be created, so that an ILogger can be used immediately within Program.cs and potentially all / any downstream components that could  benefit from logging before the container is built. LoggerFactory seems a layer on top of that I am honestly not sure if its needed - perhaps it would be better to initialise a LoggerFactory instead of just a singular ILoggerProvider. The important point for me is whichever is initialise to enable early logging, this should probably be completely seperate and independent from the host. In fact this might even just be a console app with no host and the same pattern could be used in program.cs. For example if we need an IConfiguration to initialise this logging then this can be purpose built at the point of building the LoggerFactory / ILoggerProvider -  and does not need to be conflated with the host's separate IConfiguration. The host could then be adapted to take advantage of this Logger if there is some way to provide a reference to it, so that it can log itself and pass the logger to any downstream components that would also benefit from logging such as startup classes.

**Comment by dazinator** [comment]
Updated my comment above.. my thoughts evolved!

**Comment by davidfowl** [comment]
> Logging is it's own independent subsystem represented by a LoggerFactory.

This is already the case. I'm not sure if this was a request or not.

> Templates encourage you to create / configure this logging subsystem early. Ideally by some convention.

Not sure what this means. Is it different from my proposal above? I'd like to get more concrete rather than more abstract on the suggestions 😄 .

> A Console app that is a simple CLI can use logging with same convention. To me this means without a dependency on Hosting as a CLI doesnt need to run any host, just parse and execute commands.

That's out of scope. You can call LoggerFactory.Create in a console application.

> Host / WebHost can:
leverage / depend upon / insist upon logging being set up first.

This will be opt in. I don't expect many changes to the webhost (other than potentially logging earlier if possible).


So here's another version with less things:

```C#
using System;
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;

namespace MyApplication
{
    public class Program
    {
        public static int Main(string[] args)
        {
            // This is the ring 0 logging
            using var loggerFactory = LoggerFactory.Create(builder =>
            {
                // All code based configuration, no configuration here..
                builder.AddConsole();
            });

            var logger = loggerFactory.CreateLogger<Program>();

            try
            {
                CreateWebHostBuilder(args, configuration, loggerFactory).Build().Run();
            }
            catch (Exception ex)
            {
                logger.LogError(ex, "An error occured when starting the application");
                return 1;
            }

            return 0;
        }

        public static IHostBuilder CreateWebHostBuilder(string[] args, ILoggerFactory loggerFactory) =>
            Host.CreateDefaultBuilder(args)
                .UseLoggerFactory(loggerFactory) // New method
                .ConfigureWebHostDefaults(builder =>
                {
                    builder.UseStartup<Startup>();
                });
    }
}
```

- Removes configuration, configuring "ring 0" logging requires code based configuration.
- There's still a problem with double logging. There's providers in the box by default and there's the logger factory's configured providers. If they both add console, how does that work.

**Comment by dazinator** [comment]
@davidfowl 
We are aligned, a minor thing:

> try
            {
                CreateWebHostBuilder(args, configuration, loggerFactory).Build().Run();
            }
            catch (Exception ex)
            {
                logger.LogError(ex, "An error occured when starting the application");
                return 1;
            }

Would need finally block with Disposal of LoggerFactory so loggers can flush messages  before app terminates

> There's still a problem with double logging. There's providers in the box by default and there's the logger factory's configured providers. If they both add console, how does that work.

Ring 0 logger factory will produce loggers only consumed by pre-DI activated classes. Think program.cs, hostbuilder, startup.cs etc. Classes activated after DI using the DI container that the host built will be using separate LoggerFactory. These two LoggerFactories may have the same providers depending upon how they are configured but that doesnt mean you get duplicate log messages. It means one LoggerFactory or the other will be in play and the logs will be different.

The issue remains though for console provider- if two seoerate instances are craated I assume they both wrap STDOUT and if both loggers were logging at the same time you'd end up with garbled (not duplicate) output.

I'm not sure if this is something that is unique to console log provider and could be addressed  by console log provider.

If its unique to console log provider we might  be able to put in some special casing so that this host doesnt create another instance when one is already in play?

**Comment by DamianEdwards** [comment]
I'm wondering whether it wouldn't be simpler to just standardize on the concept of "two-phase" logging: main/bootstrap/startup and hosted. Where that likely falls down is when you want to use the exact same logging configuration for both phases, but I honestly don't know how true that is, as many popular logging sinks require configuration to be initialized successfully anyway.

If we were simply to change the default template to have a startup logger that wrote to console (manually configured in the template) and then the normal default logger from within the host, how would that feel? Docs can always show more advanced scenarios like those suggested above (sharing logging configuration between main and host) but at least that default result would make it more obvious how the `ILogger` primitives can be used independently of the host to log events around it.

**Comment by dazinator** [comment]
> Where that likely falls down is when you want to use the exact same logging configuration for both phases

If the first phase consists of creating an LoggerFactory then to use the exact same configuration for the second phase (the host) should just be a very simple case of:

```
services.AddSingleton<ILoggerFactory>(firstPhaseLoggerFactoryInstance);
```

Rather than doing AddLogging() to configure a new LoggerFactory for the second phase. This has the benefit that the host container wont steal ownership of the lifetime of the phase 1 loggerfactory, and it wont result in two console logger providers.

I think I'd most often have two seperate configurations though but it's probably down do preference..

**Comment by davidfowl** [comment]
What happens when there's duplicated sinks?

**Comment by rynowak** [comment]
Have we considered doing 2 rounds of loggers via buffering?

- ring 0 logging just buffers 
- once the host is created we flush everything that was logged to the *real* logger

Is the problem with this that we're concerned about cases where the application crashes before the host is built?

**Comment by davidfowl** [comment]
> ring 0 logging just buffers
> once the host is created we flush everything that was logged to the real logger

That's another solution and what one of our customers did here https://github.com/MV10/GenericHostBuilderLogger#generichostbuilderlogger. 

> Is the problem with this that we're concerned about cases where the application crashes before the host is built?

Yes but you could imagine solving this by dumping buffered data on dispose/disposeasync. If data is buffered then it means the host never logged to begin with.

It wouldn't change the proposal much but the default configuration would avoid duplicate logging:

```C#
using System;
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;

namespace MyApplication
{
    public class Program
    {
        public static int Main(string[] args)
        {
            // This is the ring 0 logging that buffers until host creation
            using var loggerFactory = LoggerFactory.Create();

            var logger = loggerFactory.CreateLogger<Program>();

            try
            {
                CreateWebHostBuilder(args, configuration, loggerFactory).Build().Run();
            }
            catch (Exception ex)
            {
                logger.LogError(ex, "An error occured when starting the application");
                return 1;
            }

            return 0;
        }

        public static IHostBuilder CreateWebHostBuilder(string[] args, ILoggerFactory loggerFactory) =>
            Host.CreateDefaultBuilder(args)
                .UseLoggerFactory(loggerFactory) // New method
                .ConfigureWebHostDefaults(builder =>
                {
                    builder.UseStartup<Startup>();
                });
    }
}
```

**Comment by dazinator** [comment]
@davidfowl 
> What happens when there's duplicated sinks?

For some sinks it would be fine, and others I am guessing not - like console. For example, If the ring 0 LoggerFactory and the host LoggerFactory both had their own instance of the console logger provider, then, _assuming they managed to both log an event at the same time_*, you't get garbled stuff in the console window:

   * Could they both log at the same time though? ring 0 logging should happen before / and possibly after right? Maybe you could enforce a clear lifetime over the ring 0 logger factory, or prevent its usage whilst the host is running with docs?

2. This could be a matter of documenting the fact that ring 0 LoggerFactory should either be:
   - Configured with sinks to avoid duplicates with your application (or ones that work concurrently)
   - Used also for your application instead of adding a new LoggerFactory - e.g `services.AddSingleton<ILoggerFactory>(ringZeroInstance);`

**Comment by dazinator** [comment]
@davidfowl 
> What happens when there's duplicated sinks?

For some sinks it would be fine, and others I am guessing not - like console. For example, If the ring 0 LoggerFactory and the host LoggerFactory both had their own instance of the console logger provider, then, _assuming they managed to both log an event at the same time_*, you't get garbled stuff in the console window:

   * Could they both log at the same time though? ring 0 logging should happen before the host has initialised the container / and possibly after it has died right? Maybe you could enforce a clear lifetime over it's use, supplemented with documentation to prevent its usage whilst the host is running with docs - just in case there are sinks /providers that don't support concurrent usage.

This could be a matter of documenting the fact that:-
   - Some logger providers are not safe to use multiple instances concurrently (like console logger) so you should avoid creating more than one instance.
   - If you configure ring0 logging, and your main application logging individually (i.e 2 seperate logger factories) be careful not to use multiple logger providers that can't be used concurrently (like the ones covered by above docs)
   - Ideally ring0 and main appication logging shouldn't be logging at the same time. Be careful not to hang on to a ring 0 logger and use it from background tasks / threads. (If this was the case then the above problems shouldn't cause you issues anyway?)
    - Consider whether you need two seperate logger factories - perhaps you can just initialise the ring 0 logger factory and then register that instance with your application container later in startup.cs. This avoids all the issues. `services.AddSingleton<ILoggerFactory>(ringZeroInstance);`

**Comment by dazinator** [comment]
So I guess my question is: Do we want to enable different configurations for ring 0 logging and main application logging, because if we don't need that then these issues go away right?

**Comment by davidfowl** [comment]
> So I guess my question is: Do we want to enable different configurations for ring 0 logging and main application logging, because if we don't need that then these issues go away right as we can just use the one logger factory.

That goes back to the original proposal. If they are not separate then the logger needs configuration and all of the defaults that are being provided today by the host.

> P.S I don't like the buffer* idea. Some sinks like Seq, its really useful to monitor the output in real time. Flushing all the log events on shutdown can make it seem they were all generated at once when in reality there are potential timing differences between the events. It also means introducing this caching ability / behaviour into the LoggerFactory increasing it's responsibilities and complexity.

I don't see this as a problem as you would only buffer until the host is initialized. Maybe it needs to be an explicit buffering sink.

**Comment by dazinator** [comment]
> That goes back to the original proposal. If they are not separate then the logger needs configuration and all of the defaults that are being provided today by the host:

Not sure I follow. To modify your example a little bit, I've added serilog as an example; what's wrong with this:

```

using System;
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;

namespace MyApplication
{
    public class Program
    {
        public static int Main(string[] args)
        {
                 // Intialise serilog
                Log.Logger = new LoggerConfiguration()
                .Enrich.FromLogContext()
                .WriteTo.Console()
                .CreateLogger();

            // This is the ring 0 logging that buffers until host creation
            using var loggerFactory = LoggerFactory.Create((builder)=>{                 
                  builder.AddSerilog(); // etc
            });

            var logger = loggerFactory.CreateLogger<Program>();

            try
            {
                CreateWebHostBuilder(args, configuration, loggerFactory).Build().Run();
            }
            catch (Exception ex)
            {
                logger.LogError(ex, "An error occured when starting the application");
                return 1;
            }

            return 0;
        }

        public static IHostBuilder CreateWebHostBuilder(string[] args, ILoggerFactory loggerFactory) =>
            Host.CreateDefaultBuilder(args)
                .UseLoggerFactory(loggerFactory) // New method - perhaps also registers as service
                .ConfigureWebHostDefaults(builder =>
                {
                    builder.UseStartup<Startup>();
                });
    }
}

```

**Comment by dazinator** [comment]
@davidfowl 
This is a quick summary from my perspective, given everything i've read above:

- Configuration of ring 0 logging needs to be done before the Host starts, so that the host has something to use during initialisation.
- The Host is also able to configure logging (logger factory) to register into ApplicationServices - as it does today - let's call that "Ring 1" logging - to follow the ring terminology.
- The developer will want the option of having the Host re-use the Ring 0 logger factory instance as its Ring 1 logger, for passing onto the application (application services) **or** configuring an entirely new LoggerFactory for ring 1 logging as it does today.

If the latter option is used (different factory for ring 0 and ring 1) then the developer needs to worry about duplicate sink usage however:
    - Usage of ring 0 logger factory is limited in scope so that should help avoid duplicate logs - typically it will be used during startup and shutdown procedures
    - Usage of ring 1 logging works as it does today.
    - Only some sinks / providers will be susceptible to concurrency issues (such as console, where if added to seperate ring 0 and ring 1 logger factory instances, with concurrent usage, will garble messages). It's possible some additional validation on the logger factory instances providers could detect such issues in future - or warn against them, similar to your feature for how you validate DI scopes in development.

**Comment by dazinator** [comment]
@davidfowl 
.. and just a thought on how you might prevent (or validate and throw an early exception) for the duplicate sink issue:

If AddLogging builder was changed so that providers could be registered with a lifetime:

```
b=>{
 b.AddSingletonProvider(consoleProvider);
 b.AddScopedProvider(seqProvider);
}
```

Then it would become possible to check the ring 1 logger factory instances providers, get all those with a singleton lifetime, and if any of the same providers appear in the ring 0 logger factory - throw an exception. This could be an IsDevelopment only feature kind of like DI scope validation..

**Comment by nblumhardt** [comment]
Following this thread and giving staged initialization in Serilog some thought; https://github.com/nblumhardt/serilog-reload is a proof-of-concept wrapper for Serilog that switches between a synchronization-heavy mutable state and a leaner immutable state once initialization is complete.

Loggers handed out in the initial phase are lazy, caching wrappers that reload a "real" logger if the pipeline changes. The goal would be to not hand out too many of these.

The duplicate sink problem is handled by tearing down the whole initial pipeline before the new, reloaded, pipeline is created. Makes a (shaky?) assumption that the initial sinks will be cheap to create and tear down.

In the short-term, I'm exploring this to see whether we can use it for early Serilog initialization that's "frozen" when registered into the framework's container with a different set of sinks configured from `Configuration` and DI.

Might be a little O/T for this thread, but, could be one possible set of trade-offs to try; drop me a line if I can add anything. 👋

**Comment by dazinator** [comment]
Coming back to this thread a long while later...

Two competing ideas in play are:

1. Initialise ring 0 logging for use before DI is built by host. Host has potentially its own seperately configured "ring 1" logging configured as normal. Perhaps options to allow host's ring1 logging to re-use same logger provider as ring 0 logging to keep a single logging provider instance rather than 2 if a seperate ring 1 logging configuration is unnecessary.
2. Initialise ring 0 logging that simply buffers all logs, until the hosts ring 1 logging is available (after host DI is built) at which point all logs are flushed.

My concern with 2 is that I am not sure how it solves the problem should an error occur before the host is built and ring 1 logger is configured. Where will the logs be flushed too?

**Comment by dazinator** [comment]
@DamianEdwards 

> I'm wondering whether it wouldn't be simpler to just standardize on the concept of "two-phase" logging: main/bootstrap/startup and hosted. Where that likely falls down is when you want to use the exact same logging configuration for both phases, but I honestly don't know how true that is, as many popular logging sinks require configuration to be initialized successfully anyway.
> 
> If we were simply to change the default template to have a startup logger that wrote to console (manually configured in the template) and then the normal default logger from within the host, how would that feel? Docs can always show more advanced scenarios like those suggested above (sharing logging configuration between main and host) but at least that default result would make it more obvious how the `ILogger` primitives can be used independently of the host to log events around it.

I'm liking this more and more. There is already a conventon to have this CreateWebHostBuilder method on program.cs:

```

        public static IHostBuilder CreateWebHostBuilder(string[] args, ILoggerFactory loggerFactory) =>
            Host.CreateDefaultBuilder(args)             
                .ConfigureWebHostDefaults(builder =>
                {
                    builder.UseStartup<Startup>();
                });

```

So this would be extending the template to add something like:

```csharp
public static void Main(string[] args)
{
   using(var logging = LoggerFactory.Create(InitialiseLogging))
   {
        logging.Debug("Entry point called..");
       
        // Put existing stuff here with initialising host

   }
}

public static ILoggerProvider InitialiseLogging(ILoggingBuilder builder)
{         
                // maybe also look for some config file by convention? i.e logging.json?         
                builder.AddConsole();  // Note: console logger should only ever be added to one logger provider.
 }
```

This would apply to any `dotnet` project with an entry point - even those that don't initialise a host such as simple CLI applications.
